python/lsst/ctrl/bps/quantum_clustering_funcs.py

```diff
@@ -399,3 +397,191 @@ def add_cluster_dependencies(cqgraph, cluster, quantum_to_cluster):
+            ]
+        case _:
+            raise RuntimeError("Invalid find_dependency_of (method)")
+
+    new_clusters = []
+    for task_label in dim_labels:
+        task_def = qgraph.findTaskDefByLabel(task_label)
+        assert task_def is not None  # for mypy
+        for node in qgraph.getNodesForTask(task_def):
+            cluster_name, info = get_cluster_name_from_node(
+                node, cluster_dims, cluster_label, template, equal_dims
+            )
+
+            # Add cluster to the ClusteredQuantumGraph.
+            # Saving NodeId instead of number because QuantumGraph API
+            # requires it for creating per-job QuantumGraphs.
+            if cluster_name in cqgraph:
+                cluster = cqgraph.get_cluster(cluster_name)
+            else:
+                cluster = QuantaCluster(cluster_name, cluster_label, info)
+                cqgraph.add_cluster(cluster)
+            cluster.add_quantum(node.nodeId, task_label)
+
+            # Save mapping for use when creating dependencies.
+            quantum_to_cluster[node.nodeId] = cluster_name
+
+            # Use dependencies to find other quantum to add
+            # Note: in testing, using the following code was faster than
+            # using networkx descendants and ancestors functions
+            nodes_to_use = [node]
+            nodes_already_visited: dict[UUID, bool] = {}
+            while nodes_to_use:
+                node_to_use = nodes_to_use.pop()
+                match method:
+                    case "source":
+                        possible_nodes = qgraph.determineOutputsOfQuantumNode(node_to_use)
+                    case "sink":
+                        possible_nodes = qgraph.determineInputsToQuantumNode(node_to_use)
+                for possible_node in possible_nodes:
+                    if possible_node.taskDef.label in ordered_tasks[cluster_label]:
+                        if possible_node.nodeId not in nodes_already_visited:
+                            _LOG.debug(
+                                "Adding possible quantum %s (%s) to cluster %s",
+                                possible_node.nodeId,
+                                possible_node.taskDef.label,
+                                cluster_name,
+                            )
+                            cluster.add_quantum(possible_node.nodeId, possible_node.taskDef.label)
+                            quantum_to_cluster[possible_node.nodeId] = cluster_name
+                            nodes_already_visited[possible_node.nodeId] = True
+                            nodes_to_use.append(possible_node)
+                    else:
+                        _LOG.debug(
+                            "label (%s) not in ordered_tasks. Not adding possible quantum %s",
+                            possible_node.taskDef.label,
+                            possible_node.nodeId,
+                        )
```

Comment:
Was concerned about speed.  Changing to cluster.qgraph_node_ids and can optimize later if needed.

With the messages being debug ones, I wasn't worried about repeats.  If we discover this is too verbose, we can fix later.

Reply:
If you're concerned about speed, I'm fine with keeping it around. Just please rename it to something like `nodes_in_cluster` and make it a set instead of a dictionary.

Fine, but can you please add a comment that the current implementation of the graph traversal may visit some nodes repeatedly.